sources/academy/tutorials/api/retry_failed_requests.md

```diff
@@ -0,0 +1,61 @@
+---
+title: How to retry failed requests
+description: Learn how to resurrect your run but retrying only failed requests
+sidebar_position: 6
+slug: /api/retry-failed-requests
+---
+
+**Learn how to re-scrape only failed requests in your run.**
+
+---
+
+There are many reasons why requests for a scraper could fail. The most popular causes are different page layouts or proxy blocking issues ([check here on how to effectively analyze errors](https://docs.apify.com/academy/node-js/analyzing-pages-and-fixing-errors)). Both Apify and [Crawlee](https://crawlee.dev/) allow you to restart your scraper run from the point where it ended but there is no native functionality to re-scrape only failed requests. Usually, you also want to first analyze the problem, update the code and build it before trying again.
+
+If you attempt to re-run an already finished run, it will likely immediately finish because all the requests in the [request queue](https://crawlee.dev/docs/3.7/guides/request-storage) are marked as handled. So you need to update the failed requests in the queue to be marked as pending again.
+
+The additional complication is that the [Request](https://crawlee.dev/api/core/class/Request) object doesn't have anything like `isFailed` property, we have to approximate it using other fields. Fortunately, we can use the `errorMessages` and `retryCount` properties to identify failed requests. Unless the user explicitly overrode these properties, we can identify failed requests by having a larger amount of `errorMessages` than `retryCount`. That happens because the last error that doesn't cause a retry anymore is added to `errorMessages`.
+
+A simplified code example can look like this:
+
+```typescript
+// The code is similar for both Crawlee-only but uses a different API
+import { Actor } from 'apify';
+
+const REQUEST_QUEUE_ID = 'pFCvCasdvsyvyZdfD'; // Replace with your valid request queue ID
+const allRequests = [];
+let exclusiveStartId = null;
+// List all requests from the queue, we have to do it in a loop because the request queue list is paginated
+for (; ;) {
+    const { items: requests } = await Actor.apifyClient.requestQueue(REQUEST_QUEUE_ID).listRequests({ exclusiveStartId, limit: 1000 });
+    for (const request of requests) {
+        allRequests.push(requests);
+    }
+    // If we didn't get full 1000 requests, we have all and can finish the loop
```

Comment:
```suggestion
    // If we didn't get the full 1,000 requests, we have all and can finish the loop
```

Reply:
I find the 'we have all' part confusing. Is it referring to a specific status in the code? Otherwise, it doesn't sound right, but as it seems to contradict the previous clause (we didn't get vs we have all), I'm not sure what alternative to suggest.

Reply:
It goes in a loop while fetching batches of 1000 requests. If we get 1000 - it means there's more and we should continue the loop, if we have less than 1000 - it means that it was the last batch and we got the leftovers, so nothing else to fetch.

Reply:
So the meaning here is something like:
// If we didn't get the full 1,000 requests of the current batch, we have fetched all the requests from the queue and can finish the loop